sklearn/manifold/_locally_linear.py

```diff
@@ -119,7 +119,11 @@ def barycenter_kneighbors_graph(X, n_neighbors, reg=1e-3, n_jobs=None):
     ind = knn.kneighbors(X, return_distance=False)[:, 1:]
     data = barycenter_weights(X, X, ind, reg=reg)
     indptr = np.arange(0, n_samples * n_neighbors + 1, n_neighbors)
-    return csr_matrix((data.ravel(), ind.ravel(), indptr), shape=(n_samples, n_samples))
+    return csr_matrix(
+        (data.ravel(), ind.ravel(), indptr),
+        shape=(n_samples, n_samples),
+        dtype=X.dtype,
```

Comment:
There is no need to force the `dtype` here because `data` returned by `barycenter_weights` is preserving the `dtype`.